bigtable_automation/gcf/bt_trigger.go

```diff
@@ -6,311 +20,237 @@
-		return "", "", status.Errorf(codes.Internal, "Unexpected cache type %s", parts[0])
+	nodesL, err := strconv.Atoi(nodesLow)
+	if err != nil {
+		return err
 	}
-	return parts[0], parts[1], nil
-}
-
-func btImportControllerInternal(ctx context.Context, e GCSEvent) error {
-
-	env := envs[CurrentEnv]
-
-	if e.Bucket != env.controlBucket {
+	// Get control bucket and object
+	controlBucket, _, err := parsePath(controlPath)
+	if err != nil {
+		return err
+	}
+	if e.Bucket != controlBucket {
 		return status.Errorf(codes.Internal, "Unexpected bucket %s", e.Bucket)
 	}
+	// Get table ID.
+	parts := strings.Split(e.Name, "/")
+	tableID := parts[len(parts)-2]
 
 	if strings.HasSuffix(e.Name, initFile) {
 		log.Printf("[%s] State Init", e.Name)
 		// Called when the state-machine is at Init. Logic below moves it to Launched state.
-
-		// (base|branch)/<tableID>/<initFile>
-		cacheType, tableID, err := parsePath(e.Name)
+		launchedPath := fmt.Sprintf("%s/%s/%s", controlPath, tableID, launchedFile)
+		exist, err := isObjectExist(ctx, launchedPath)
 		if err != nil {
-			return err
-		}
-		btInstance := ""
-		if cacheType == baseCacheType {
-			btInstance = env.baseBTInstance
-		} else {
-			btInstance = env.branchBTInstance
+			return errors.WithMessagef(err, "Failed to check %s", launchedFile)
 		}
-		launchedPath := fmt.Sprintf("%s/%s/%s", cacheType, tableID, launchedFile)
-		if _, err := readFromGCS(ctx, e.Bucket, launchedPath); err == nil {
-			return status.Errorf(codes.Internal, "Cache was already built for %s/%s: %v", cacheType, tableID, err)
+		if exist {
+			return errors.WithMessagef(err, "Cache was already built for %s", tableID)
 		}
-		if err := setupBTTable(ctx, env.btProjectID, btInstance, tableID); err != nil {
+		if err := setupBT(ctx, projectID, instance, tableID); err != nil {
 			return err
 		}
-		if cacheType == baseCacheType {
-			// Scale up only for base cache.
-			if err := scaleBT(ctx, env.btProjectID, env.baseBTInstance, env.baseBTClusters, env.baseBTNodesHigh); err != nil {
-				return err
-			}
+		if err := scaleBT(ctx, projectID, instance, cluster, int32(nodesH)); err != nil {
+			return err
 		}
-		err = launchDataflowJob(ctx, env, btInstance, cacheType, tableID)
+		err = launchDataflowJob(ctx, projectID, instance, tableID, dataPath, controlPath, dataflowTemplate)
 		if err != nil {
 			return err
 		}
 		// Save the fact that we've launched the dataflow job.
-		err = writeToGCS(ctx, e.Bucket, launchedPath, "")
+		err = writeToGCS(ctx, launchedPath, "")
 		if err != nil {
 			return err
 		}
 		log.Printf("[%s] State Launched", e.Name)
 	} else if strings.HasSuffix(e.Name, completedFile) {
 		log.Printf("[%s] State Completed", e.Name)
 		// Called when the state-machine moves to Completed state from Launched.
-
-		// (base|branch)/<tableID>/<completedFile>
-		cacheType, _, err := parsePath(e.Name)
-		if err != nil {
+		if err := scaleBT(ctx, projectID, instance, cluster, int32(nodesL)); err != nil {
```

Comment:
If I read this correctly, we are now also scaling up branch BT instance?

Then we probably should make sure the total of nodesHigh between branch + base is under 300 (the quota).  So perhaps we should update nodesHigh for branch

Reply:
Added additional check to resolve https://b.corp.google.com/issues/218700670 as well. Now it only scales up when the target node number is higher than current node number. And only scales down if the target nodeHigh is same as the current node number